core/server/src/main/java/alluxio/worker/file/UnderFileSystemManager.java

```diff
@@ -0,0 +1,439 @@
+   * manager.
+   */
+  public UnderFileSystemManager() {
+    mIdGenerator = new AtomicLong(IdUtils.getRandomNonNegativeLong());
+    mInputStreamAgents =
+        new IndexedSet<>(mInputStreamAgentSessionIdIndex, mInputStreamAgentIdIndex);
+    mOutputStreamAgents =
+        new IndexedSet<>(mOuputStreamAgentSessionIdIndex, mOutputStreamAgentIdIndex);
+  }
+
+  /**
+   * An object which generates input streams to under file system files given a position. This
+   * class does not manage the life cycles of the generated streams and it is up to the caller to
+   * close the streams when appropriate.
+   */
+  @ThreadSafe
+  private final class InputStreamAgent {
+    /** Session this agent belongs to, used as an index. */
+    private final long mSessionId;
+    /** Id referencing this agent, used as an index. */
+    private final long mAgentId;
+    /** Configuration to use for this stream. */
+    private final Configuration mConfiguration;
+    /** The string form of the uri to the file in the under file system. */
+    private final String mUri;
+
+    /**
+     * Constructor for an input stream agent for a UFS file. The file must exist when this is
+     * called.
+     *
+     * @param sessionId the session this agent belongs to
+     * @param agentId the agentId of the agent
+     * @param ufsUri the uri of the file in the UFS
+     * @param conf the configuration to use
+     * @throws FileDoesNotExistException if the file does not exist in the UFS
+     * @throws IOException if an error occurs when interacting with the UFS
+     */
+    private InputStreamAgent(long sessionId, long agentId, AlluxioURI ufsUri, Configuration conf)
+        throws FileDoesNotExistException, IOException {
+      mSessionId = sessionId;
+      mAgentId = agentId;
+      mUri = ufsUri.toString();
+      mConfiguration = conf;
+      UnderFileSystem ufs = UnderFileSystem.get(mUri, mConfiguration);
+      ufs.connectFromWorker(conf,
+          NetworkAddressUtils.getConnectHost(NetworkAddressUtils.ServiceType.WORKER_RPC, conf));
+      if (!ufs.exists(mUri)) {
+        throw new FileDoesNotExistException(
+            ExceptionMessage.UFS_PATH_DOES_NOT_EXIST.getMessage(mUri));
+      }
+    }
+
+    /**
+     * Opens a new input stream to the file this agent references. The new stream will be at the
+     * specified position when it is returned to the caller. The caller is responsible for
+     * closing the stream.
+     *
+     * @param position the absolute position in the file to start the stream at
+     * @return an input stream to the file starting at the specified position
+     * @throws IOException if an error occurs when interacting with the UFS
+     */
+    private InputStream openAtPosition(long position) throws IOException {
+      UnderFileSystem ufs = UnderFileSystem.get(mUri, mConfiguration);
+      InputStream stream = ufs.open(mUri);
+      if (position != stream.skip(position)) {
```

Comment:
why is skip used instead of seek? The default implementation of InputStream.java uses read to do skip. That is expensive.